### ISSUE TITLE
[filters] Backdrop filters should not use BackgroundImage

### ISSUE THREAD

**Issue by grorg** [w3c/fxtf-drafts#53]
Currently, the spec says "The first filter function or <a element>filter</a> reference in the list takes the element's BackgroundImage as the input image."

This was a mistake on my part. BackgroundImage is defined as the backdrop up to the isolation parent. 

What we really want is a new term, BackdropImage, which is the backdrop of the element, including all parents and siblings that have been painted up to the point of the filter. There is no isolation.

**Comment by AmeliaBR** [comment]
Is there going to be a BackdropAlpha or any other way to interpret the backdrop? If not, I think "backdrop" is both shorter and less likely to be confused with CSS background images.

But while we're on the subject:

Why no isolation? How would this interact with mix-blend-modes in general?

What do you mean "parents and sibling that have been painted"? Is that, after proper re-ordering for z-index, or only previous DOM siblings?

How do you determine which section of the backdrop is affected by the filter? Is it clipped to the current element's content/border/margin-box?  Is it masked by the current element's painted alpha channel?

**Comment by grorg** [comment]
Many great questions!

> Why no isolation? How would this interact with mix-blend-modes in general?

It's not that it isn't isolated, just that the backdrop isn't clamped at the isolation root.

The mix-blend-mode affects how this element is then composited. So it is effectively rendered offscreen and then blended. Backdrop filters won't change that.

> What do you mean "parents and sibling that have been painted"? Is that, after proper re-ordering for z-index, or only previous DOM siblings?

I meant in paint order, not DOM order. I just couldn't think of an existing term for that.

The way backdrop filter works is that you do a separate rendering pass just for the backdrop. This is why it is expensive.

> How do you determine which section of the backdrop is affected by the filter? Is it clipped to the current element's content/border/margin-box? Is it masked by the current element's painted alpha channel?

At the moment it is the border box. It isn't masked by the element's painted alpha. Maybe we'll need an option to do that for SVG elements.

**Comment by mstange** [comment]
Can you say more about why you chose to make backdrop filter not respect isolation? I had always assumed they respected isolation (i.e. just operated on the implied intermediate surface of the nearest stacking context, the way blend modes do), and was quite surprised when I played with them in Safari a few weeks ago and found out that this was not the case. I also noticed very inconsistent behavior, see below. Do you have scenarios in mind where respecting isolation would impose unnecessary limitations on what authors could do?

If you're inside a stacking context and render a backdrop filter, does the separate rendering pass include the contents of the current stacking context up to the current element? This testcase shows rather contradictory behavior in Safari:

```html
<div style="height: 200px; opacity: 0.5; background: black;">
  <div style="height: 100px; -webkit-backdrop-filter: opacity(1);"></div>
</div>
```

On initial load, I just get a 200px high gray box (i.e. white + 50% black). If I then use the Web Inspector to change `opacity(1)` to `opacity(0)` and back to `opacity(1)`, suddenly the upper half of the box becomes white. What's the intended behavior here?

**Comment by mstange** [comment]
I've filed the problem from the previous comment as [https://bugs.webkit.org/show_bug.cgi?id=164606](https://bugs.webkit.org/show_bug.cgi?id=164606).

**Comment by mstange** [comment]
@grorg This issue definitely needs clarification before we can agree to implementing backdrop-filter in Firefox.

In particular, it needs to be clear how the testcase in https://github.com/w3c/fxtf-drafts/issues/53#issuecomment-250811306 should behave.

There are strong parallels to `mix-blend-mode` here: If you have an element with `mix-blend-mode`, blending needs to happen between that element's contents and the element's backdrop. We have an existing definition for what exactly the `mix-blend-mode` backdrop is, and that definition respects isolation. If `backdrop-filter` ends up using a different definition for what the backdrop is, there should be a good justification for the difference.

**Comment by dirkschulze** [comment]
@AmeliaBR @mstange From looking at the example, nothing escapes isolation as far as I can see.

1. Backdrop filter takes the entire content of the backdrop. 
2. Everything drawn up to the backdrop filter respects the `isolation` property.
3. The white circle is drawn on top of the element that is drawn with the backdrop filter (and the actual backdrop)
4. Difference filter is still isolated to the common group of the element with the backdrop filter and the white circle with difference.

I understand that it sounds strange but isolation rules are still met for blending. The one is independent of the other.

As @grorg says, we need a new term to describe this behavior.
* `BackgroundImage` and `BackgroundAlpha` stop at isolation. Nothing before the isolation gets used.
* `Backdrop` wouldn't stop at isolation but the rest of the drawing (`filter`, `mix-blend-mode`) would still respect isolation.

It would be interesting to see if Safari runs into issues where it normally would create a compositing layer. So if the group would have an animation of the `transform` property for instance. Or a 3D transform gets applied. This usually complicates things for implementations. I would be impressed if both still work. If not, we probably need separate definitions of `isolation`.

**Comment by dirkschulze** [comment]
Run 2 different tests:
* One with 2D transform animation
* One with 3D transform animation and perspective.

## 2D tranform animation
https://codepen.io/krit/pen/XqXvyo

* Once an animation with a transform gets specified and starts, the group starts getting isolated. So the 2nd circle in the first example gets white on hover.
* `backdrop-blur` continues to get the content from the backdrop.

This actually makes sense. Even when split into compositing layers, the implementation still has access to the different "raster images"/compositing layers and can composite those for the backdrop blur filter. At least when you implement backdrop-blur in the paint pipeline you have.

## 3D transform animation with perspective
https://codepen.io/krit/pen/zjrgmy
Then I modified the example with a `perspective` property and a `rotateX` (I always struggle to make things look 3D but it works for this test 😆 ).

* On hover, the `backdrop-filter` stops working. This is expected since we no longer have two 2D planes that simply needs to get composited.

## Conclusion
The question is:
* Do we make `backdrop-filter` stop at `isolation` as `filter` and `mix-blend-mode` do or
* do we define a new isolation term based on the existence of perspective?

And extended question:
* If we go with the new isolation term, can we have it for `BackgroundImage` too?

**Comment by mfreed7** [comment]
@grorg, this definitely needs a few clarifications before we can get it shipped in Chrome. The main sticking point is whether `backdrop-filter` respects isolation. Our reading of the spec agrees with your interpretation - it states that the filter takes the `BackgroundImage` as its input, and that is defined ([here](https://drafts.fxtf.org/filter-effects-1/#attr-valuedef-in-backgroundimage))
to be the contents of the current isolation group up to the current element. For several reasons, we like that definition:

* Making backdrop-filter respect isolation boundaries would make it behave in exactly the same way that other filter effects do, including `mix-blend-mode` and `filter` (with `in=’BackgroundImage’`). This should reduce developer confusion, as compared to introducing another brand-new concept specifically for use with backdrop-filter.

* If backdrop-filter is re-defined to filter everything behind the element, including items outside the parent stacking context, the implementation gets much more technically involved. Two complete rendering passes need to be performed, one to build the image “behind” the element as input to the filter, and the other to actually render it all together. This additional complexity can lead to more bugs and performance degradation.

We believe that having backdrop-filter respect isolation boundaries will not pose a practical limitation to developers, assuming they understand the behavior correctly. For example, if they nest the popup window (with backdrop-filter) directly within the `<html>` node, it will filter everything painted before it. The only real issue would be deeply-nested elements with backdrop-filters.
 
In addition to the isolation question, as @AmeliaBR pointed out, several other things are unclear at the moment. In particular, the spec does not describe which portion of the backdrop is affected by the filter. This needs to be defined, both in terms of border (content/border/margin) and whether the current element’s painted alpha masks the filter. It seems to us like it should be clipped to the border (respecting border radius), but painted alpha shouldn’t mask the effect. This is the most straightforward, both to understand and to implement.

**Comment by mfreed7** [comment]
Another issue with defining `backdrop-filter` to read back everything (not respecting isolation) is that there are some situations where the behavior is ambiguous or at least not easily anticipated. For example, if an element (call it “A”) has 0.5 opacity (creating a stacking context), and a child of that element (call it “B”) has a backdrop-filter applied, the 0.5 opacity will potentially be “applied twice” if the isolation is not respected. First, during the read back of “everything before” the backdrop-filter on B, the 0.5 opacity will be applied to A to blend it with its backdrop (call that “R” for root). The resulting backdrop-filtered image will then be used as the starting canvas for element B. Once B is painted, this entire stacking context (containing A, the backdrop-filtered image, and B) will be composited into R using the 0.5 opacity for the group. So the portion of A under B will effectively have opacity 0.25.
 
Please see [this example](https://jsbin.com/gihilun/edit?html,output) for this effect possibly happening, in Box 3. I’ve included screenshots from Chrome, Safari, and Edge below. In Chrome, all three boxes seem to have the right amount of blur and opacity applied. (Don’t worry, there are plenty of other bugs in Chrome’s backdrop-filter, just not here.) In Safari, Box 3 looks like it is at 0.25 opacity - double-counted. And incidentally, Box 1 also shows another bug (I believe) in Safari. It has a background color, which seems to turn off the backdrop-filter entirely. On Edge, I’m not sure what is happening. The edges of the red box are crisp all the way around, so perhaps the red background is being repainted on top of the blur? Or maybe some form of opacity double-counting here too?
 
Just to make the point again - having backdrop-filter respect isolation should lead to fewer bugs, and a less confusing experience for developers.

Chrome 69:
<img width="581" alt="chrome_screenshot" src="https://user-images.githubusercontent.com/29559695/46891309-4c585c00-ce1e-11e8-91b2-9016fd2fd4fa.png">

Safari 12:
![safari_screenshot](https://user-images.githubusercontent.com/29559695/46891317-54180080-ce1e-11e8-8adc-1ecb72153595.png)

Edge 42:
![edge_screenshot_zoomed](https://user-images.githubusercontent.com/29559695/46891325-59754b00-ce1e-11e8-917d-e6653328130f.png)

**Comment by Benimation** [comment]
I think it's most important that this will work the same way it does in Safari. In my opinion, it would be more confusing for developers if it didn't work the same everywhere. Do note that the same effect is present in iOS, where it does work similarly to the way it works in Safari (if not exactly the same way).

I really do think that the backdrop should include everything behind an element, as it's supposed to be a filter that can be put on top of anything (and filter anything behind it). If it doesn't work this way, it won't look good in countless situations.

**Comment by mfreed7** [comment]
@Benimation, I definitely agree that we need to clarify the standard and then make sure the behavior is the same on all browsers.

Can you give an example of the effect "not looking good" with backdrop filter stopping at the parent stacking context? It would be helpful to see what the limitations would actually be.

**Comment by chrishtr** [comment]
Note that the [codepen](https://codepen.io/krit/pen/pxOMdz) posted by @dirkschulze has a typo in it (isolation: isolated instead of isolation: isolate).

If that is [fixed](https://codepen.io/anon/pen/ReEobj) then shows that indeed mix-blend-mode only mixes with the containing isolated group, not all the way to the root.

So the semantics of backdrop-filter that @mfreed7 is proposing is consistent with mix-blend-mode.

**Comment by mfreed7** [comment]
I just wanted to quickly add a comment about the limitations of stopping backdrop-filter at an isolation boundary. It seems like there may be some misconception about how that limits what gets filtered. Please see [this jsbin](https://jsbin.com/viqolif/2/edit?html,output), which shows a dialog element floating over the top of (and filtering) all other elements on the page. The dialog is first in the DOM hierarchy, and the other elements include several nested stacking contexts. The point is that the stacking context limitation only applies if the backdrop-filtered element itself is nested inside a stacking context for some reason. As long as it sits outside, everything behind it on the page, including other stacking contexts, gets filtered. See also, https://www.apple.com - they use exactly this DOM structure for the nav bar at the top. I'd be very interested to see examples where the backdrop filtered element **needs** to be inside a stacking context for some reason - i.e. what type of effect isn't possible with the proposed isolation behavior?

![jsbin example](https://user-images.githubusercontent.com/29559695/47677281-ef79d700-db7b-11e8-91b2-27e4a2d80a0d.png)

**Comment by mfreed7** [comment]
To see how people are using backdrop-filter today, I did some research using [HTTP Archive](https://httparchive.org/) and [Chrome Platform Stats](https://www.chromestatus.com/metrics/css/timeline/popularity/508). The current usage rate for backdrop-filter (unprefixed) is very low, about 0.0003%. That number does not include prefixed (-webkit-backdrop-filter) usage, so I used HTTP Archive to pull all pages that use backdrop-filter. Of those, about half use **only** the prefixed version, so the actual usage rate might be higher by 2-5X, which is still quite small. Another analysis based on summed Alexa ranks for each page results in a factor of about 3x. In all cases, the existing usage is quite small.

I also looked at the pages themselves that currently make use of backdrop-filter. I was only able to locate two high-ranking sites that have a visible element styled with backdrop-filter. One is [Apple](https://www.apple.com/), and the other is [the Github Blog](https://blog.github.com/). Both use nearly the same DOM structure, as I mentioned above: an element nested directly within the `<body>` with a high z-index value. And as I mentioned, this works perfectly well with our proposal to have backdrop-filter respect isolation boundaries.

We are still curious what effects would be difficult to achieve, if backdrop-filter were specified to respect isolation boundaries. Input appreciated!

**Comment by mstange** [comment]
I took a closer look at [the example](https://jsbin.com/gihilun/edit?html,output) from https://github.com/w3c/fxtf-drafts/issues/53#issuecomment-429447287 and found a potential problem with making `backdrop-filter` respect isolation. [Here's a variation of the example](https://jsbin.com/cexoheduwi/edit?html,output) with the following changes: The opacity on the ancestor element `.colorbox` was removed, and the opacity on `.box1` and `.box3` was reduced to `0.1`. This variation of the example no longer demonstrates what happens when an ancestor item is isolated, but it demonstrates what happens when opacity and backdrop-filter are applied to the same element. Here's how Chrome and Safari render it:

Chrome 72 Canary:
<img width="658" alt="screen shot 2018-11-07 at 7 41 28 pm" src="https://user-images.githubusercontent.com/961291/48170387-59555780-e2c5-11e8-9dc9-eb627d761d73.png">

Safari 12:
![screen shot 2018-11-07 at 7 41 31 pm](https://user-images.githubusercontent.com/961291/48170393-5ce8de80-e2c5-11e8-8654-c8031571b4b7.png)

In Chrome, it seems that the opacity on the element does not reduce the opacity of what's rendered by backdrop-filter on that element; the blur in the example fully blocks out the non-blurred element underneath. In Safari, the blurred backdrop is only rendered at 0.1 opacity. I think Safari's behavior makes more sense here. 

However, if we want to adopt Safari's behavior for this case, in terms of the compositing effect graph, I think this requires rendering the backdrop filter *inside* the opacity, if opacity and backdrop-filter are applied to the same element. But if the backdrop filter rendering goes inside of the opacity, then we have the same problem as in this issue: Opacity is an isolation boundary, so backdrop-filter on such an element would not have any effect. @mfreed7, any ideas for how we should solve this case?

**Comment by AmeliaBR** [comment]
> I was only able to locate two high-ranking sites that have a visible element styled with backdrop-filter.

Instagram uses it for their "please log in" banner when you follow a link directly to a photo page.  It's nested many divs deep & some of those divs use `position: relative`, but the banner itself is `position: fixed`, so for isolation I guess that count's as a direct child of the window?  (In general, looking only for "visible elements" with the style may undercount usage, because a common use case is for pop-ups and other dynamic content.)

**Comment by mfreed7** [comment]
@AmeliaBR, thanks for the comment, I hadn't noticed the Instagram photo page in my search. The numbers I quoted were for any document containing "backdrop-filter", including CSS or embedded style within an HTML. The platform stats numbers are counted any time the feature is actually used, e.g. when the pop-up pops up. So I think my numbers are still good; correct me if I'm wrong. My subsequent hand-analysis was to just use document.querySelectorAll() for the selector containing backdrop-filter and see what comes back, and it is definitely possible I missed some cases there. Again, thanks for the pointer to Instagram. From the sound of it, though, the Instagram banner would filter everything on the page, given that its parent stacking context is the root element. I am assuming none of the position:relative parents use z-index. Let me know if you disagree. (And if you could post a link directly to a photo from Instagram, that'd be helpful too.)

@mstange, thanks also for the comment about mixing opacity and backdrop-filter. It is definitely an interesting question, though I don't think it actually affects the isolation question. It does call into question the order-of-operations for backdrop-filter. We are debating it now, and I'll have a more detailed response soon.

**Comment by mfreed7** [comment]
@mstange, I definitely understand the point you’re trying to make - that the results of the backdrop-filter operation are not being painted along with the element itself, including other effects applied to the element (opacity in this case). And I agree that the spec is not very clear on the order of operations for other effects applied to the same element. However, it is important to note that this question is (I think) entirely separate from the question of whether backdrop-filter respects isolation boundaries. The isolation question merely affects which elements get included in the background image that forms the starting point for the backdrop filter. Orthogonal to that, there is the question of how to process that background image, including your question about opacity.

Turning to the specific question of filtering and backdrop-filtering on the same element, I would submit that there are many such ambiguities that really need to be nailed down in this spec. The “right” way to specify and implement it is a grey area. For example, your description assumes that the result of the backdrop-filter should be used as the starting image within the new stacking context created by the backdrop filter. Under that definition, you’re right, the opacity should be applied to the filtered backdrop. But the spec actually doesn’t mention exactly what the behavior is supposed to be. We can see good cases to be made on both sides of this question.

@chrishtr and I currently think that the behavior **should** be defined to be something like the steps listed below. I believe this clearly defines the order of operations, including filters and transforms, and it also clears up some other ambiguities about the clip border for the backdrop. In addition, and importantly, it is relatively straightforward to implement this in a performant manner. Your feedback is appreciated:

1. Draw the ancestor stacking context (call it A), in paint order, up to the backdrop-filtered element (call it B). Assume the buffer being drawn to this point is called A’.
2. Copy everything painted so far into a temporary buffer (call it T’), and apply the backdrop-filter’s filter operations to that temporary buffer.
3. Apply a clip to the contents of T’, using the border box of element B. Note that the children of B are not considered for this clip.
4. Draw all of element B and the children of B, including any transforms and other effects, into T’.
5. Composite the contents of T’ into A’, using source-over compositing.
6. Continue drawing the remainder of the children of A.

Without a definition such as the above, it is our belief that many interactions (such as your opacity example) will be difficult for developers to reason about, and will also lead to implementation issues and performance problems. With this definition, the behavior is clear and the implementation is straightforward. It also does not change the order of operations that currently apply to stacking contexts. Notably, this does not address your initial point/example - opacity applied to B will **not** be applied to the backdrop-filtered backdrop. To achieve that, however, you could just change your filter spec to <code>backdrop-filter: blur(10px) opacity(0.1);</code>. That would apply the opacity to the filtered content, and should do exactly what you expect.

Thoughts appreciated.

**Comment by smfr** [comment]
I'm still strongly in favor of having backdrop-filter affect all of the content behind, not just the enclosing isolation group. We want [this example](https://codepen.io/smfr/pen/EdeqNK) to blur the striped body background:

**Comment by mfreed7** [comment]
@smfr, thanks for the comment. Interestingly, due to a bug in Chrome, that example already works. I believe the <code>z-index:0</code> isn’t properly creating a stacking context.  However, in the general case, can you please give more information about **why** you want that example to work? I.e. by moving one div tag, you could make it work in either case. What **can’t** you do, if backdrop-filter respects isolation?

**Comment by Benimation** [comment]
Right now there's a big difference between Safari and Chrome on the homepage of apple.com

![apple-homepage-safari-12](https://user-images.githubusercontent.com/5503862/48300388-b840e980-e4dc-11e8-9d4f-168cf644240c.jpg)

![apple-homepage-chrome-72](https://user-images.githubusercontent.com/5503862/48300389-bb3bda00-e4dc-11e8-88dd-8f4be6a5f089.jpg)

The filtering appears to be extending beyond the actual navigation bar, and on the left and right the edges are letting some white color through.

**Comment by mfreed7** [comment]
@Benimation, thanks. Those are known bugs in the Chrome implementation. See https://crbug.com/659501 and https://crbug.com/767997, for example. The "extension beyond the nav bar" issue is actually because this feature is under-specified - see step #3 in my list above.

**Comment by mstange** [comment]
In response to @mfreed7:

> The isolation question merely affects which elements get included in the background image that forms the starting point for the backdrop filter. Orthogonal to that, there is the question of how to process that background image, including your question about opacity.

That's true. I was conflating the questions "where does the input come from" and "where does the output go", because I was thinking about this in terms of an implementation that would render based on an intermediate representation (some kind of effect graph) that has lost information about this distinction.

> Without a definition such as the above, it is our belief that many interactions (such as your opacity example) will be difficult for developers to reason about, and will also lead to implementation issues and performance problems.

I fully agree. We need a definition with this level of precision in the spec.

> opacity applied to B will **not** be applied to the backdrop-filtered backdrop. To achieve that, however, you could just change your filter spec to `backdrop-filter: blur(10px) opacity(0.1);`. That would apply the opacity to the filtered content, and should do exactly what you expect.

Well, mostly. It would not do group opacity the way I expect: `<div style="opacity: 0.8;backdrop-filter: blur(10px) opacity(0.8);"><div style="width:100px;height:100px;background:white"></div></div>` - in this example, you'd still be able to see remains of the blurred backdrop behind the white square, whereas the effect I'd like to achieve would have the white square block out the blur entirely. Moreover, having to repeat the opacity value in two different properties makes opacity animations a bit more cumbersome.

**Comment by mstange** [comment]
Coming back to original issue:

If we decide that backdrop-filter should respect isolation, then the web author is in control of the isolation boundaries. If we decide that it should *ignore* isolation, then the web author cannot choose an isolation boundary even if they want isolation.

There has been a proposal to allow web authors to turn off the implied isolation on stacking contexts: `isolation: accumulate`. This value [used to be part of compositing-2](https://github.com/w3c/fxtf-drafts/blob/a7fda53a9bbe527c025f0ead13e53498511d07fd/compositing/index.html) but was removed again. [This section about `enable-background: accumulate` in an old SVG spec](https://www.w3.org/TR/2009/WD-SVGCompositing-20090430/#enable-background) has a relatively precise description of how it would work.

**Comment by smfr** [comment]
> @smfr, thanks for the comment. Interestingly, due to a bug in Chrome, that example already works. I believe the `z-index:0` isn’t properly creating a stacking context. However, in the general case, can you please give more information about **why** you want that example to work? I.e. by moving one div tag, you could make it work in either case. What **can’t** you do, if backdrop-filter respects isolation?

I'm not sure what you mean by "by moving one div tag". As a web author, I should not have to rejigger which elements on my page create stacking context in order to get the effect that everything behind my element with the backdrop gets filtered, and in fact that may not be possible (maybe my isolation ancestor is transformed, but I also want backdrop filter to include non-transformed elements).

**Comment by tabatkins** [comment]
> I'm not sure what you mean by "by moving one div tag".

Moving the thing that wants a whole-page filtered backdrop to be higher in the DOM structure.

> As a web author, I should not have to rejigger which elements on my page create stacking context in order to get the effect that everything behind my element with the backdrop gets filtered

This is indeed already a requirement for a number of things in CSS; stacking contexts are very definitely **not** no-ops for a number of things. Why is backdrop-filtering, in particular, special in this regard, such that it should ignore the "layering" rules that several other related CSS features adhere to?

**Comment by mfreed7** [comment]
@tabatkins, in particular mix-blend-mode, which seems to my eye to be very similar to backdrop-filter in this regard. I don’t know why the behavior there (mix-blend-mode respects isolation) shouldn’t be the same behavior here, for backdrop-filter.

**Comment by tabatkins** [comment]
I'm a little confused - your comment seems to be in support of my point? Is that right?

**Comment by mfreed7** [comment]
@tabatkins yes, sorry - I'm supporting your point.

**Comment by smfr** [comment]
Here's another test case I think should work: https://codepen.io/smfr/pen/aQJxzM. An ancestor has a transform (making it a stacking context) but you want backdrop-filter to affect things outside it.

**Comment by smfr** [comment]
> @tabatkins, in particular mix-blend-mode, which seems to my eye to be very similar to backdrop-filter in this regard. I don’t know why the behavior there (mix-blend-mode respects isolation) shouldn’t be the same behavior here, for backdrop-filter.

Historically I think mix-blend-mode should have affected everything behind it, but for ease of implementation it also chose to stop at the isolation group ancestor. @cabanier probably remembers more.

**Comment by mfreed7** [comment]
@smfr, good example. You would have to do [this instead](https://codepen.io/mfreed7/pen/mQWgLW), which was slightly trickier than your example, but perfectly do-able.

I think it's important to mention - we're not claiming that our proposal isn't a limitation. **It is**. There will be some things that cannot be done, if backdrop-filter respects isolation. But in exchange for that limitation, (all) developers will get two benefits. The first is that the performance of sites using backdrop-filter will be much better - exponentially so, in some cases. The second is that the filter behavior will be more easily understood, so implementations will more likely match each other, and developers will know what to expect. And most "simple" backdrop-filter effects will work just the way you would expect - they would filter everything behind them. See, e.g. apple.com and blog.github.com.

I think it's important to use mix-blend-mode as a model here, because it is so similar in operation to backdrop-filter. The decision was made to have mix-blend-mode respect isolation, and that decision should be the perfect precedent here.

**Comment by mfreed7** [comment]
> Well, mostly. It would not do group opacity the way I expect: `<div style="opacity: 0.8;backdrop-filter: blur(10px) opacity(0.8);"><div style="width:100px;height:100px;background:white"></div></div>` - in this example, you'd still be able to see remains of the blurred backdrop behind the white square, whereas the effect I'd like to achieve would have the white square block out the blur entirely. Moreover, having to repeat the opacity value in two different properties makes opacity animations a bit more cumbersome.

@mstange I agree with the above. It is both more cumbersome, and as you point out, the white square will "show through" some of the blurred backdrop. If you didn't want that, you'd have to move the white square out from under the parent with opacity. Depending on what effect you wanted, you could likely achieve it, however. It would just take more work.

**Comment by chrishtr** [comment]
After further research, prototyping and testing, we need one more thing. In addition to backdrop-filter filtering only the containing isolated group, I think we need that the containing isolated group to be
a containing block for all descendants (with quirk for root). Detailed reasoning below. It is unfortunate, but I think necessary, that this change be made. Unfortunate because it is the first example I know of where style on a descendant element affects stacking behavior of an ancestor.

Note that for the, as far as we know, highly dominant use case of backdrop-filter within the root stacking context, the change I am proposing has no effect.

Reasoning:

The reason is the same as that for CSS filter (see #11):
a) Ensure that filters, especially those that move pixels, be applied atomically and not have clips or scrolls that escape them (*)
b) Ensures that the GPU accelerated rendering path can apply all effects in a hierarchical way

(a) ensures that we don't have any ill-defined cases where clips and filters don't commute, and both (a) and (b) ensure that we can avoid any slow paths in the GPU acceleration of the browser. In particular, we must ensure that composited scrolling is not unduly slowed down by backdrop-filter.

(*) Example to avoid: a position:{fixed/absolute} element which is a child of the containing isolated group to backdrop-filter. Without making the isolated group a containing block for such elements, this element can skip clips and scrolls above the backdrop-filter in the DOM, which is illogical.

**Comment by dirkschulze** [comment]
Just a note @chrishtr, CSS Transforms require a containing block for all descendants already. So filter wouldn’t be the first.

**Comment by mfreed7** [comment]
@dirkschulze, the difference here is that the **ancestor** stacking context of the backdrop-filtered element would need to be made a containing block. Not the element itself, unfortunately.

**Comment by mfreed7** [comment]
@mstange, we were discussing your [comment above](https://github.com/w3c/fxtf-drafts/issues/53#issuecomment-438431485) concerning opacity. There are several examples where it would make sense for the filtered backdrop to form the starting point for the backdrop-filtered element, so that opacity (and other filters) would apply to that filtered content. One example would be a dialog box that “fades out”. Given that this change should not cause any performance issues, I’ve modified our proposed definition for backdrop-filter. I also included more detail about transforms applied to the backdrop-filtered element, to clarify those as well. Your feedback is appreciated:

1. Draw the ancestor stacking context (call it A), in paint order, up to the backdrop-filtered element (call it B). Assume the buffer being drawn to this point is called A’.
2. Copy everything painted so far into a temporary buffer (call it T’), and apply the backdrop-filter’s filter operations to that temporary buffer.
3. If element B has any transforms, apply the **inverse** of those transforms to the contents of T’.
4. Apply a clip to the contents of T’, using the border box of element B. Note that the children of B are not considered for this clip.
5. Draw all of element B and the children of B into T’.
6. If element B has any transforms, effects, or clips, apply those to T’.
7. Composite the contents of T’ into A’, using source-over compositing.
8. Continue drawing the remainder of the children of A.

Again, we believe that having backdrop-filter only apply up to the parent stacking context makes sense for these reasons:
1. It is the only definition that can be unambiguous for the case of nested elements with filters and opacity applied. If we simply define it to filter “everything behind”, there will be **many** corner cases that will need to be very carefully and explicitly defined; otherwise, each implementation will inevitably produce different output. See, e.g., [my original example](https://github.com/w3c/fxtf-drafts/issues/53#issuecomment-429447287) with three different behaviors from three browsers, for just the simple case of opacity applied to the parent element.
2. It will allow performant implementations. As far as we can tell, nested backdrop-filters that don’t respect isolation will suffer exponential performance degradation, as each one doubles the required rendering work.
3. It is exactly analogous to <code>mix-blend-mode</code>, and we see no reason to have different behavior for similar operations. In addition, all of the logic that went into making mix-blend-mode respect isolation should also apply to backdrop-filter.

**Comment by mfreed7** [comment]
We would like to publish a proposal for spec changes for backdrop-filter. It includes many of the ideas discussed on this issue, plus a lot of extra background and motivation for the changes. I have created a PR for the fxtf-drafts repo, and I've included the rendered spec output here:

https://mfreed7.github.io/fxtf-drafts/filter-effects-2/Overview.html

We would love to hear comments on this proposal.

@ChumpChief, thanks for your comments. In our proposed spec, I've included a lot of detail in the "Motivation" section. In particular, I added a rationale for each of the restrictions included in the definition of the Backdrop Root. Hopefully that's a good starting point for the discussion!

**Comment by AmeliaBR** [comment]
@mfreed7 Is this proposal a description of what one of the browsers currently does, or would it require some changes for everyone?

**Comment by mfreed7** [comment]
Good question, @AmeliaBR. Only one browser currently ships backdrop-filter: Edge. Webkit has a prefixed version, and Mozilla and Chromium don't have a shipped backdrop-filter implementation.
In terms of what's actually implemented in code, behind flags:
* We have been working towards this proposal in our implementation over the last several months in Chromium, so we are likely closest to this description at this point. We still have some work to do.
* Webkit's and Edge's implementations mostly differ in that they do not stop filtering at filter, opacity, or mask/clip nodes.
* I do not know what is implemented in Mozilla, if anything. (@mstange has previously sounded supportive of some parts of this proposal - I would love to hear his thoughts here.)

**Comment by mstange** [comment]
> * I do not know what is implemented in Mozilla, if anything. (@mstange has previously sounded supportive of some parts of this proposal - I would love to hear his thoughts here.)

Mozilla has not started implementing backdrop-filter, mostly because of the lack of clarity on this very issue. (And also due to resource constraints.)

I've taken a look at the proposal, and I think it's great! Really terrific work. It clarifies all the previously-unspecified points (at least the ones I can think of), and it addresses the use cases that have been brought up before which would have been limited by the previously-proposed approach of respecting isolation.
However, I do have a few concerns:

  - I don't believe making the backdrop root a containing block is a good idea. More on that below.
  - I would like to see some discussion of how the `isolation` property plays into this. I see that `isolation: isolate` is not among the properties that cause backdrop roots. Should it be?
  - Making rounded-corner overflow clips cause backdrop roots seems like an unnecessary restriction to me. In Firefox, we make no attempt to apply the rounded corner clip to the entire contents of the clipped element in one step; instead, the rounded clip can be pushed down into the "leaves" of the effect tree whenever necessary, and we accept the imperfect anti-aliasing (and the possibility of bleeding) that this creates. And in some cases, applying the rounded overflow clip to the entire contents isn't even possible, and the clip *has* to be applied to different parts of the element's contents individually. (I'm referring to the interleaved `<roundedclip><A/><fixedposition/><B/></roundedclip>` case.) This is an existing scenario that browsers have to support, and I think in Firefox, our existing solution to this would just work for the purposes of computing the backdrop-blur input.
  - Speaking of Firefox's implementation, our compositor currently has two more triggers for intermediate surfaces that are not on the list of backdrop root triggers:

      - Clipped transforms (even if the clip is just a rectangle overflow clip without rounded corners): If you have `<rectangleclip><rotation>layerized contents</rotation></rectangleclip>`, then our compositor creates an intermediate surface for the rotation and applies the rectangle clip when drawing the rotated surface into the outer render target.
      - 3D transforms, even unclipped, always get an intermediate surface.

    I don't think we should add these two cases to the list of backdrop root triggers, because it would defeat much of the point, but this does mean that, at least in our initial implementation, Firefox will likely need to do an extra render pass to compute the backdrop input inside transforms, and won't be able to just use the contents of the current render target.

Now, to my main objection: The creation of a containing block for the backdrop root. Quoting from the proposed text:

> Note: In addition to the stacking context and containing block formed by the element with backdrop-filter, the corresponding Backdrop Root element also forms a Containing Block for all descendants. This is to avoid child absolute or fixed position elements that do not get clipped by parent clips or overflow:scroll elements, but which need to be read back to form the Backdrop Root Image.

I think this is not a good idea for two reasons: 1. It has non-local effects and 2. it doesn't seem necessary to me.

Here's an example to demonstrate the first point:

```html
<div style="opacity: 0.9; height: 100px;">
  <div style="position: fixed; bottom: 0; background: green; width: 20px; height: 20px;"></div>
  <div style="backdrop-filter: blur(1px)"></div>
</div>
```

As far as I can tell, the mere presence of the element with the `backdrop-filter` changes the position of the fixed element: If the `backdrop-filter` element is *not* present, then the fixed element will be fixed with respect to the page and will be located at the bottom of the browser viewport. But if the `backdrop-filter` element *is* present, then the wrapper element with the opacity becomes a backdrop root and suddenly turns into the containing block for the fixed element, and the fixed element ends up in a completely different place. I think this behavior would be very surprising to web authors.
It would also mean that the answer to the question "Is element A a containing block for fixed-position descendants?" depends on element A's *descendants*, which is very unusual.

I realize that this point was brought up in November by @chrishtr on this issue.

> (*) Example to avoid: a position:{fixed/absolute} element which is a child of the containing isolated group to backdrop-filter. Without making the isolated group a containing block for such elements, this element can skip clips and scrolls above the backdrop-filter in the DOM, which is illogical.

I agree it's not intuitive, but I think we're kinda stuck with the behavior of elements escaping clips and scrolling while still being subject to effects from DOM ancestors. Browsers already have to support nesting such as `<scrollbox><opacity><A/><fixed/><B/></opacity></scrollbox>`; in this case the opacity group contains both the fixed content and the non-fixed content. It seems to me that taking this "mixed content" as the input to a `backdrop-filter` is just as hard as applying opacity to it. Or in other words, I don't see the benefit from a new restriction for `backdrop-filter` since the existing infrastructure for this case should be able to handle it just fine.

**Comment by AmeliaBR** [comment]
Would it be possible to extend `isolation` to have a third value that means "don't isolate, even if this is a stacking context or other element that usually causes `auto` isolation".  This would be similar to how [`transform-style: preserve-3d`](https://drafts.csswg.org/css-transforms-2/#transform-style-property) works.

(Currently, `isolation` only has two values, `isolate` vs `auto`.)

This would address the authoring issue (authors often expect the backdrop filter to filter everything behind the element, but maybe they also expect isolation to limit it if they're using blending or opacity). But it could be more complicated for implementers, as they would need to implement both behaviors, both for backdrop filtering and for blend modes.

**Comment by mfreed7** [comment]
I have updated my pull request per the discussion just now:
* Removed overflow:clip with rounded corners as a Backdrop Root trigger
* Removed the containing block induction note - Backdrop Root no longer is a containing block.
* Added an Issue link at the top, currently pointing to this issue, but can be redirected to a new issue.

Link is (still) here: https://mfreed7.github.io/fxtf-drafts/filter-effects-2/Overview.html

**Comment by mfreed7** [comment]
The new spec is located at the "normal" place now: https://drafts.fxtf.org/filter-effects-2/

**Comment by mfreed7** [comment]
Ugh. The new spec was written as it was because nesting backdrop filter within elements that inherit (like filters and opacity) causes fundamental problems. *Please* see the Example under [Backdrop Root Motivation](https://drafts.fxtf.org/filter-effects-2/#BackdropRootMotivation), or the [example above of when things go wrong, in Webkit](https://github.com/w3c/fxtf-drafts/issues/53#issuecomment-451599226), because of this very problem.

**Comment by mstange** [comment]
FWIW, nothing new was discussed during today's meeting. Please don't read too much into the log. From what I can tell, every browser maker's position remains the same as in March.

(I can clarify my comment about opacity but I don't think it's relevant.)

**Comment by mfreed7** [comment]
Thanks @mstange. I think I understand what you're going for on the opacity comment - basically, you're saying that "filtering everything" will require special rules for what to do in the myriad nesting cases for opacity (and also filters). Unless the spec is basically "just put something on the screen there, I don't care what it is", it'll be tough to spec, I suspect.

**Comment by tabatkins** [comment]
Yeah, the core idea is that unless/until WK actually describes their behavior, which they were tasked with doing at the *last* meeting, we're going with the current spec (Chrome) behavior. And if they don't do that reasonably quickly, Firefox will likely follow the spec, and begin fully locking in that behavior.

**Comment by mfreed7** [comment]
Thanks @tabatkins. Do you think July 30, 2019 (the M76 stable release date) constitutes "reasonably quickly"?

**Comment by tabatkins** [comment]
That's up to Firefox, so I dunno.

**Comment by cabanier** [comment]
Can we update the CSS Composition spec so it refers to backdrop root instead of stacking context?

**Comment by mfreed7** [comment]
@cabanier, I would love that change, but don't you think it would cause web compat issues? The [use counter](https://chromestatus.com/metrics/css/timeline/popularity/420) is fairly high, and the visual appearance might change if mix-blend-mode starts blending more of the backdrop.

**Comment by AmeliaBR** [comment]
The compat issue against changing `mix-blend-mode` could be addressed by adding a new [`isolation` value](https://drafts.fxtf.org/compositing-2/#propdef-isolation). `auto` (the default) could still isolate by stacking context, but there could be an override (`dont-isolate`?) that uses the full backdrop, if there are no technical reasons not to support that.

Connecting that back to `backdrop-filter`: the `auto` value of isolation is defined to be sensitive to a lot of context.  If the default isolation behavior creates nonsensical results when combined with `backdrop-filter`, then a different behavior could be defined specifically for the interaction between the two properties.

**Comment by cabanier** [comment]
@mfreed7 yeah, there could be content that will suddenly look different. I'm unsure if this could be detected by a use counter; it could be too costly to calculate.

@AmeliaBR in one of the previous versions of the spec, we had the `non-isolated` property (which matches the Adobe products). Maybe we can bring that back? 
An unfortunate drawback is that there will still be content that will always force isolation (ie filters). I'm unsure how confusing this is for users.

**Comment by samer-Sp** [comment]
> Currently, the spec says "The first filter function or filter reference in the list takes the element's BackgroundImage as the input image."
> This was a mistake on my part. BackgroundImage is defined as the backdrop up to the isolation parent.
> What we really want is a new term, BackdropImage, which is the backdrop of the element, including all parents and siblings that have been painted up to the point of the filter. There is no isolation.

**Comment by Mampinator** [comment]
Look, I am not an expert on how visuals are set up on the browser at all, but I am a web developer, and understand just how broken simple things such as being able to mask or have opacity for items with backdrop filters are in Chromium. It is just flat out broken right now, and it has been since Sept. 29 2016 apparently. That is crazy. 

I know there has been a lot of debate regarding spec and how to implement it, but as a starter, I will just say that other browser such as FireFox seem to have implemented it just fine. So if you are having trouble figuring out what to do, analyze the behavior in their browser. It does everything I would expect the backdrop-filter to do at a minimum, and that standard is unfortunately already very low. The web needs some visual upgrades in CSS.

Bottom line is that there needs to be more time put into this. It is not just a small issue because it is not a security issue. **There are not many visual tools on the web in CSS. The least you can do is make the existing ones work...**

**Comment by Mampinator** [comment]
I am bringing this issue back to the present because I have seen that the thread has died down and still, it is not closed. This issue needs to be fixed.

**Comment by Mampinator** [comment]
Here is a Chromium ticket that is being neglected: https://issues.chromium.org/issues/334110494. It goes hand in hand with the issue that has been being discussed on this thread for **8 years.** 😳

**Comment by MrChriZ** [comment]
Agreed this is nuts.  My site with backdrop filter looks great in firefox... but it won't work in the world's leading browser because the spec doesn't allow for it?!

**Comment by greg-wolff** [comment]
**9 years later...........**

**Comment by aaronventure** [comment]
Bump

**Comment by GanerCodes** [comment]
<img width="801" height="803" alt="Image" src="https://github.com/user-attachments/assets/2be9176f-931d-4bf9-aee0-c7db80600291" />
